File: iPacMG/cookie_login.pl
#!/usr/bin/perl

## This page checks to see if a user is logged in.
## If they are not logged in, they are presented with a login form.
## simply: include 'cookie_login.pl';
## Nothing more is necessary.
## This page is meant to run SQL Queries as-well. (since it's cookie protected)
## For example, two variables are already set: "$real_username" and "$real_email"
## So any page that has this page "included" can use $real_username and get the user's username... Ect…
## Go ahead and set more variables by fetching more stuff from DB

use CGI;
use MIME::Base64;
use DBI();
use DBI;
use CGI::Cookie;
use HTTP::Cookies;

BEGIN{
    $q = new CGI;
    {   $method = $q->param("method");
        if($method =~ "set"){
                $read_u_cookie = $q->param("u_cookie");
                $read_p_cookie = $q->param("p_cookie");
                $u_cookie = new CGI::Cookie(-name=>'u_cookie',-value=>$read_u_cookie,-path=>'/');
            	$p_cookie = new CGI::Cookie(-name=>'p_cookie',-value=>$read_p_cookie,-path=>'/');
                $url = "cookie_login.pl";
                print $q->header(-Location=>$url,-cookie=>[$u_cookie, $p_cookie],-type=>'text/html', -charset => 'UTF-8');
                
        } else {
#print $q->header(-type=>'text/html', -charset => 'UTF-8');
        }
    }
    open(STDERR, ">&STDOUT");
}

$method = $q->param("method");

%cookies = CGI::Cookie->fetch;
$got_u_cookie = $cookies{'u_cookie'}->value;
$got_p_cookie = $cookies{'p_cookie'}->value;


if(!defined $got_u_cookie && !defined $got_p_cookie){
print "<form action='n_login.php' method='post'>\n";
print "Username: <input type='text' name='username'/><br>\n";
print "Password: <input type='password' name='password'/><br>\n";
print "<input type='hidden' name='method' value='login'/>\n";
print "<input type='submit' value='Login'/><br>\n";
print "</form>\n";
exit;
}
$uc_safe =~ s/'//g;
$uc_safe =~ s/"//g; # php version adds slashes, perl removes the ', " in all.
$pc_safe =~ s/'//g;
$pc_safe =~ s/"//g;

# $dbh, & $sth have _random123164 extension bc/ they seem to trail into other files...

my $dbh_random123164 = DBI->connect("DBI:mysql:database=shortlando;host=localhost",
"shortland", "password!",
{'RaiseError' => 1});

my $sth_random123164 = $dbh_random123164->prepare("SELECT u_cookie, p_cookie, username, email, valid FROM users WHERE u_cookie = ? AND p_cookie = ?");
$sth_random123164->execute($got_u_cookie, $got_p_cookie) or die "Couldn't execute statement: $DBI::errstr; stopped";

while ( my ($u_cookie, $p_cookie, $username, $email, $user_valid) = $sth_random123164->fetchrow_array() ) {
  $real_username = decode_base64($username);
  $real_email = $email;
  $real_access = $user_valid;
  $valid = "ok";
}
$dbh_random123164->disconnect();

if (!defined $valid){
# Invalid, old, currupt cookies
# Force user to login again.
# print "error is in validation";
print "<script>window.location.href='n_login.php';</script>";
print "<form action='n_login.php' method='post'>\n";
print "Username: <input type='text' name='username'/><br>\n";
print "Password: <input type='password' name='password'/><br>\n";
print "<input type='hidden' name='method' value='login'/>\n";
print "<input type='submit' value='Login'/><br>\n";
print "</form>\n";
exit;
}

1;

File: iPacMG[released][server]/cookie_login.php
<?php
// This page checks to see if a user is logged in.
// If they are not logged in, they are presented with a login form.
// simply: include 'cookie_login.php';
// Nothing more is necessary.
// This page is meant to run SQL Queries as-well. (since it's cookie protected)
// For example, two variables are already set: "$real_username" and "$real_email"
// So any page that has this page "included" can use $real_username and get the user's username... Ect...
// Go ahead and set more variables by fetching more stuff from DB


echo "

<meta name='viewport' content='height = device-height, width = device-width, initial-scale = 1, minimum-scale = 1, maximum-scale = 1, user-scalable = no, target-densityDpi=device-dpi' />

<style>
* {
-webkit-appearance: none;
}
#resp {
color: red;
}
#text {
width: 120px;
height: 22px;
border: 1px solid black;
border-radius:5px;
font-weight: bold;
}
#submit {
width: 80px;
height:30px;
border: 1px solid black;
color: black;
background-color: white;
border-radius: 5px;
}
</style>";


if(!isset($_COOKIE['u_cookie']) && !isset($_COOKIE['p_cookie'])){
echo "
<script>
  if (top.location.href != self.location.href) {
     top.location.href = self.location.href;
    }
</script>
";

echo "<style>

#topbb {
position:fixed;
top:0px;
left:0px;
height:20px;
width:100%;
background-color:lightblue;
border-bottom:1px solid black;
z-index:999999;
}

</style>";

echo "<span id='topbb'></span>";
echo "<BR><BR><BR>";
echo "<center>";
echo "<form action='n_login.php' method='POST'>\n";
echo "<input id='text' type='hidden' name='username'/><br><br>\n";
echo "<input id='text' type='hidden' name='password'/><br><br>\n";
echo "<input type='hidden' name='method' value='login'/>\n";
echo "<input id='submit' type='submit' value='Next'/><br>\n";
echo "</form>\n";
echo "</center>";
die();
}
$uc_safe = addslashes($_COOKIE['u_cookie']);
$pc_safe = addslashes($_COOKIE['p_cookie']);
$connect = mysqli_connect("localhost","shortland","password!","shortlando");
if (mysqli_connect_errno()) {
  echo "CANT CONNECT:" . mysqli_connect_error();
}
$check = mysqli_query($connect, "SELECT `u_cookie`, `p_cookie`, `username`, `email` FROM `users` WHERE `u_cookie` = '$uc_safe' AND `p_cookie` = '$pc_safe'") or die(mysqli_error($connect));
// Get the salt by username, use salt to hash the password. then check if new hashed password is == to the stored password.
while($row = mysqli_fetch_array($check)) {
  // use these variables in files that have THIS file included. no need to do mysql queries to get username ect.
  $real_username = base64_decode($row['username']);
  $real_email = $row['email'];
  $valid = "ok";
}
if (!isset($valid)){
// Invalid, old, currupt cookies
// Force user to login again.


echo "
<script>
  if (top.location.href != self.location.href) {
     top.location.href = self.location.href;
    }
</script>
";

echo "<style>

#topbb {
position:fixed;
top:0px;
left:0px;
height:20px;
width:100%;
background-color:lightblue;
border-bottom:1px solid black;
z-index:999999;
}

</style>";

echo "<span id='topbb'></span>";
echo "<BR><BR><BR>";
echo "<center>";
echo "<form action='n_login.php' method='POST'>\n";
echo "<input id='hidden' type='hidden' name='username'/><br><br>\n";
echo "<input id='hidden' type='hidden' name='password'/><br><br>\n";
echo "<input type='hidden' name='method' value='login'/>\n";
echo "<input id='submit' type='submit' value='Next'/><br>\n";
echo "</form>\n";
echo "</center>";
die();
}
?>